File: Beam/Source/QueriesTests/RangedQueryTester.cpp
#include <sstream>
#include <doctest/doctest.h>
#include "Beam/Queries/RangedQuery.hpp"
#include "Beam/SerializationTests/ValueShuttleTests.hpp"

using namespace Beam;
using namespace Beam::Tests;

TEST_SUITE("RangedQuery") {
  TEST_CASE("default_constructor") {
    auto query = RangedQuery();
    REQUIRE(query.get_range() == Range::EMPTY);
  }

  TEST_CASE("range_constructor") {
    auto total_range_query = RangedQuery(Range::TOTAL);
    REQUIRE(total_range_query.get_range() == Range::TOTAL);
    auto real_time_range_query = RangedQuery(Range::REAL_TIME);
    REQUIRE(real_time_range_query.get_range() == Range::REAL_TIME);
    auto sequence_range_query = RangedQuery(Range(Sequence(1), Sequence(2)));
    REQUIRE(
      sequence_range_query.get_range() == Range(Sequence(1), Sequence(2)));
  }

  TEST_CASE("set_range") {
    auto query = RangedQuery();
    REQUIRE(query.get_range() != Range::TOTAL);
    query.set_range(Range::TOTAL);
    REQUIRE(query.get_range() == Range::TOTAL);
    query.set_range(Range::REAL_TIME);
    REQUIRE(query.get_range() == Range::REAL_TIME);
    query.set_range(Sequence(1), Sequence(2));
    REQUIRE(query.get_range() == Range(Sequence(1), Sequence(2)));
  }

  TEST_CASE("stream") {
    auto query = RangedQuery(Range::TOTAL);
    auto ss = std::stringstream();
    ss << query;
    REQUIRE(ss.str() == "Total");
    test_round_trip_shuttle(query);
  }
}

File: WebApi/tests/queries/ranged_query_tester.ts
import { describe, it } from 'node:test';
import * as assert from 'node:assert';
import { Range } from '../../source/queries/range';
import { RangedQuery } from '../../source/queries/ranged_query';
import { Sequence } from '../../source/queries/sequence';

describe('RangedQuery', () => {
  it('default_constructor', () => {
    assert.ok(new RangedQuery().range.equals(Range.EMPTY));
  });

  it('range', () => {
    const query = new RangedQuery();
    query.range = Range.TOTAL;
    assert.ok(query.range.equals(Range.TOTAL));
  });

  it('equals', () => {
    assert.ok(new RangedQuery(Range.TOTAL).equals(new RangedQuery(Range.TOTAL)));
    assert.ok(!new RangedQuery(Range.TOTAL).equals(new RangedQuery()));
  });

  it('round_trip', () => {
    const query =
      new RangedQuery(new Range(new Sequence(1n), new Sequence(9n)));
    const json = query.toJson();
    assert.deepStrictEqual(json.range, query.range.toJson());
    assert.ok(RangedQuery.fromJson(json).equals(query));
  });
});
